qiskit_metal/analyses/hamiltonian/states_energies.py

```diff
@@ -0,0 +1,115 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+This file contains functions related to extracting and manipualting Hamiltonians
+and their energy levels.
+"""
+from typing import Dict as Dict_
+from typing import List
+
+import numpy as np
+import qutip
+from qutip import Qobj
+
+
+def basis_state_on(mode_size: List[int], excitations: Dict_[int, int]):
+    ''' Construct a qutip Qobj. Taken from pyEPR
+    https://github.com/zlatko-minev/pyEPR/blob/master/pyEPR/calcs/back_box_numeric.py
+
+    Args:
+        mode_size (list): list of integers specifying number of fock states
+                           for each mode, respectively
+        excitations (dict): {mode index: # of photons}
+            give the value excitations[i] or 0 if index i not in excitations;
+            i.e., by default ground state.
+
+    Returns:
+        qutip Qobj: qutip tensor product representing the state
+    '''
+    return qutip.tensor(*[
+        qutip.basis(mode_size[i], excitations.get(i, 0))
+        for i in range(len(mode_size))
+    ])
+
+
+def extract_energies(esys_array: np.ndarray,
+                     mode_size: List[int],
+                     zero_evals: bool = True,
+                     chi_prime: bool = False):
+    """
+    Returns the frequencies, anharmonicities, and dispersive shifts of the modes.
+
+    Args:
+        esys_array (np.ndarray): numpy array of shape (2, ). It's an array of objects.
+            The first element of the array is an array of eigenvalues of the diagonalized
+            hamiltonian. The second element of the array is an QutipEigenStates object,
+            which is a list of the corresponding eigenstates of the diagonalized hamiltonian
+        mode_size (List[int]): list of integers specifying number of fock states
+            for each mode, respectively
+        zero_evals (bool, optional): If true, the "ground state" eigenvalue is substracted
+            all eigenvalues. Defaults to True.
+        chi_prime (bool, optional): Defaults to False.
+
+    Returns:
+        np.ndarray, np.ndarray: a tuple of arrays. The first array is the frequencies of
+            the modes. The second array is a matrix where the diagonal entries are the
+            anharmonicities of the modes and the off-diagonal entries are the dispersive
+            shifts, i.e., the chi's between the modes
+    """
+
+    print("Processing eigensystem...", end='')
+    evals, evecs = esys_array
+    print("\rFinished eigensystem.     ")
+
+    if zero_evals:
+        evals -= evals[0]  # zero out
+
+    # if use_1st_order:
+    # TOOD: Copy from pyEPR
+    def closest_state_to(s: Qobj):
+
+        def distance(s2: Qobj):
+            return (s.dag() * s2[1]).norm()
+
+        return max(zip(evals, evecs), key=distance)
+
+    N = len(mode_size)
+
+    def state_on(excitations):
+        return basis_state_on(mode_size, excitations)  # eigenstate on
+
```

Comment:
Since the variable names are short, could you provide some more information.

i.e. fs, ev, d, f1s